python/ray/util/client/server/server.py

```diff
@@ -113,24 +117,34 @@ def _return_debug_cluster_info(self, request, context=None) -> str:
         return json.dumps(data)
 
     def release(self, client_id: str, id: bytes) -> bool:
-        if client_id in self.object_refs:
-            if id in self.object_refs[client_id]:
-                logger.debug(f"Releasing object {id.hex()} for {client_id}")
-                del self.object_refs[client_id][id]
-                return True
-
-        if client_id in self.actor_owners:
-            if id in self.actor_owners[client_id]:
-                logger.debug(f"Releasing actor {id.hex()} for {client_id}")
-                del self.actor_refs[id]
-                self.actor_owners[client_id].remove(id)
-                return True
-
-        return False
+        with self.state_lock:
+            if client_id in self.object_refs:
+                if id in self.object_refs[client_id]:
+                    logger.debug(
+                        f"Releasing object {id.hex()} for {client_id}")
+                    del self.object_refs[client_id][id]
+                    return True
+
+            if client_id in self.actor_owners:
+                if id in self.actor_owners[client_id]:
+                    logger.debug(f"Releasing actor {id.hex()} for {client_id}")
+                    self.actor_owners[client_id].remove(id)
+                    if self._can_remove_actor_ref(id):
+                        logger.debug(f"Deleting reference to actor {id.hex()}")
+                        del self.actor_refs[id]
+                    return True
+
+            return False
 
     def release_all(self, client_id):
```

Comment:
Can you add a unit test for the refcounting changes?

Reply:
done in separate pr